native/src/main/java/io/ballerina/stdlib/graphql/runtime/engine/ListenerUtils.java

```diff
@@ -18,25 +18,114 @@
 
 package io.ballerina.stdlib.graphql.runtime.engine;
 
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.values.BArray;
 import io.ballerina.runtime.api.values.BObject;
+import io.ballerina.runtime.api.values.BString;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static io.ballerina.stdlib.graphql.runtime.utils.Utils.ERROR_TYPE;
+import static io.ballerina.stdlib.graphql.runtime.utils.Utils.createError;
 
 /**
  * External utility methods used in Ballerina GraphQL listener.
  */
 public final class ListenerUtils {
     private static final String HTTP_SERVICE = "graphql.http.service";
+    private static final String GRAPHIQL_SERVICE = "graphql.graphiql.service";
+
+    private static final String SAMPLE_URL = "http://localhost:9000/";
+    public static final String GRAPHIQL_RESOURCE = "graphiql.html";
+    private static final String REGEX_URL = "${url}";
+    private static final String FORWARD_SLASH = "/";
 
     private ListenerUtils() {}
 
     public static void attachHttpServiceToGraphqlService(BObject graphqlService, BObject httpService) {
         graphqlService.addNativeData(HTTP_SERVICE, httpService);
     }
 
+    public static void attachGraphiQLServiceToGraphqlService(BObject graphqlService, BObject httpService) {
+        graphqlService.addNativeData(GRAPHIQL_SERVICE, httpService);
+    }
+
     public static Object getHttpServiceFromGraphqlService(BObject graphqlService) {
         Object httpService = graphqlService.getNativeData(HTTP_SERVICE);
         if (httpService instanceof BObject) {
             return httpService;
         }
         return null;
     }
+
+    public static Object getGraphiQLServiceFromGraphqlService(BObject graphqlService) {
+        Object graphiqlService = graphqlService.getNativeData(GRAPHIQL_SERVICE);
+        if (graphiqlService instanceof BObject) {
+            return graphiqlService;
+        }
+        return null;
+    }
+
+    public static Object validateGraphiQLPath(BString path) {
+        String uri = SAMPLE_URL + path;
+        try {
+            new URL(uri).toURI();
+            return null;
+        } catch (URISyntaxException e) {
+            return createError("Invalid path provided for GraphiQL client", ERROR_TYPE);
+        } catch (MalformedURLException e) {
+            return createError("Invalid path provided for GraphiQL client", ERROR_TYPE);
+        }
```

Comment:
Since we are returning the same error, we don't need two separate catch blocks, right?

I'd also suggest to add the message from the exception for a clearer error.

```suggestion
        } catch (URISyntaxException | MalformedURLException e) {
            return createError("Invalid path provided for GraphiQL client", ERROR_TYPE);
        }
```